src/app/qgsmaptooltrimextendfeature.cpp

```diff
@@ -0,0 +1,285 @@
+    {
+      if ( layer )
+        return match.layer() == layer && match.hasEdge();
+
+      return match.hasEdge();
+    }
+    // We only want to modify the current layer. When geometries are overlapped, this makes it possible to snap onto the current layer.
+    void setLayer( QgsVectorLayer *vlayer ) { layer = vlayer; }
+
+  private:
+    const QgsVectorLayer *layer = nullptr;
+};
+
+QgsMapToolTrimExtendFeature::QgsMapToolTrimExtendFeature( QgsMapCanvas *canvas )
+  : QgsMapToolEdit( canvas )
+{
+  mToolName = tr( "Trim/Extend feature" );
+}
+
+QgsMapToolTrimExtendFeature::~QgsMapToolTrimExtendFeature()
+{
+}
+
+static void getPoints( const QgsPointLocator::Match &match, QgsPoint &p1, QgsPoint &p2 )
+{
+  const QgsFeatureId fid = match.featureId();
+  const int vertex = match.vertexIndex();
+
+  const QgsGeometry geom = match.layer()->getFeature( fid ).geometry();
+
+  if ( !( geom.isNull() || geom.isEmpty() ) )
+  {
+    p1 = geom.vertexAt( vertex );
+    p2 = geom.vertexAt( vertex + 1 );
+  }
+}
+
+void QgsMapToolTrimExtendFeature::canvasMoveEvent( QgsMapMouseEvent *e )
+{
+  mapPoint = e->mapPoint();
+
+  FeatureFilter filter;
+  QgsPointLocator::Match match;
+
+  switch ( step )
+  {
+    case 0:
+
+      match = mCanvas->snappingUtils()->snapToMap( mapPoint, &filter );
+      if ( match.isValid() )
+      {
+        is3DLayer = QgsWkbTypes::hasZ( match.layer()->wkbType() );
+
+        QgsPointXY p1, p2;
+        match.edgePoints( p1, p2 );
+
+        mRubberBandLimit.reset( createRubberBand( QgsWkbTypes::LineGeometry ) );
+        mRubberBandLimit->addPoint( p1 );
+        mRubberBandLimit->addPoint( p2 );
+        mRubberBandLimit->show();
+
+      }
+      else
+      {
+        if ( mRubberBandLimit )
+          mRubberBandLimit->hide();
+      }
+      break;
+    case 1:
+
+      QgsMapLayer *currentLayer = mCanvas->currentLayer();
+      if ( !currentLayer )
+        break;
+
+      vlayer = qobject_cast<QgsVectorLayer *>( currentLayer );
+      if ( !vlayer )
+        break;
+
+      if ( !vlayer->isEditable() )
+        break;
+
+      filter.setLayer( vlayer );
+      match = mCanvas->snappingUtils()->snapToMap( mapPoint, &filter );
+
+      if ( match.isValid() )
+      {
+        if ( match.layer() != vlayer )
+          break;
+
+        QgsPointXY p1, p2;
+        match.edgePoints( p1, p2 );
+
+        getPoints( match, pExtend1, pExtend2 );
+
+        // No need to trim/extend if segments are continuous
+        if ( ( ( pLimit1 == pExtend1 ) || ( pLimit1 == pExtend2 ) ) || ( ( pLimit2 == pExtend1 ) || ( pLimit2 == pExtend2 ) ) )
+          break;
+
+        segmentIntersects = QgsGeometryUtils::segmentIntersection( pLimit1, pLimit2, pExtend1, pExtend2, intersection, isIntersection, 1e-8, true );
+
+        if ( is3DLayer && QgsWkbTypes::hasZ( match.layer()->wkbType() ) )
+        {
+          /* Z Interpolation */
+          QgsLineString line( QVector<QgsPoint>()
```

Comment:
There's a direct two-points constructor for QgsLineString which would be more efficient to use.